src/DocumentFormat.OpenXml/Packaging/RelationshipErrorHandler.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.IO.Packaging;
+using System.Xml.Linq;
+
+namespace DocumentFormat.OpenXml.Packaging
+{
+    /// <summary>
+    /// A class that holds information about invalid relationships in <see cref="PackagePart"/> instances.
+    /// </summary>
+    public abstract class RelationshipErrorHandler
+    {
+        private const string RelationshipsTagName = "Relationships";
+        private const string RelationshipTagName = "Relationship";
+        private const string RelationshipNamespaceUri = "http://schemas.openxmlformats.org/package/2006/relationships";
+        private const string TargetAttributeName = "Target";
+        private const string IdAttributeName = "Id";
+
+        /// <summary>
+        /// Rewrites an invalid URI with a valid one in order to correctly open a package.
+        /// </summary>
+        /// <param name="partUri">Uri of the part with the invalid relationship.</param>
+        /// <param name="id">Id of relationship</param>
+        /// <param name="uri">Invalid <see cref="Uri"/></param>
+        public abstract string Rewrite(Uri partUri, string id, string uri);
+
+        /// <summary>
+        /// Callback for after a package has been completely loaded and all rewritting has occurred.
+        /// </summary>
+        public virtual void OnPackageLoaded()
+        {
+        }
+
+        internal void Handle(Package package)
+        {
+            if (package is null)
+            {
+                throw new ArgumentNullException(nameof(package));
+            }
+
+            if ((package.FileOpenAccess & FileAccess.ReadWrite) != FileAccess.ReadWrite)
+            {
+                throw new OpenXmlPackageException(ExceptionMessages.PackageAccessModeIsReadonly);
+            }
+
+            foreach (var part in package.GetParts())
+            {
+                if (part.Uri.OriginalString.EndsWith(".rels", StringComparison.OrdinalIgnoreCase))
+                {
+                    var doc = WalkRelationships(part);
+
+                    if (doc != null)
+                    {
+                        using var stream = part.GetStream(FileMode.Open, FileAccess.Write);
+
+                        doc.Save(stream);
```

Comment:
I am puzzled about a question: If we open a document in read-only mode, will this code change the document?

Reply:
No, it will not. It will throw. If you want it to remain read-only, you'd need to operate on a copy.

Reply:
@twsouthwick If we follow this idea, does our arguments lack the necessary information to support the method of copying file copies? For example, I want to copy files in this code, but I don’t know which file is currently being used.

Reply:
Can you give an example of what you mean? I'm not sure I follow. From what I understand, you'd do this

```csharp
using var fs = File.OpenRead(....);
using var ms = new MemoryStream();

fs.CopyTo(ms);

using var doc = WordDocument.Open(ms);
```

Reply:
@twsouthwick Thank you. This method seems to solve the problem. But does this code take up too much memory?

Reply:
Potentially, at that point, you'd want to offload it to a file. Depends on the file